core/src/either.rs

```diff
@@ -56,99 +57,75 @@ where
 
 /// Implements `AsyncRead` and `AsyncWrite` and dispatches all method calls to
 /// either `First` or `Second`.
+#[pin_project]
 #[derive(Debug, Copy, Clone)]
 pub enum EitherOutput<A, B> {
-    First(A),
-    Second(B),
+    First(#[pin] A),
+    Second(#[pin] B),
 }
 
 impl<A, B> AsyncRead for EitherOutput<A, B>
 where
-    A: AsyncRead + Unpin,
-    B: AsyncRead + Unpin,
+    A: AsyncRead,
+    B: AsyncRead,
 {
-    fn poll_read(mut self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, IoError>> {
-        match &mut *self {
-            EitherOutput::First(a) => AsyncRead::poll_read(Pin::new(a), cx, buf),
-            EitherOutput::Second(b) => AsyncRead::poll_read(Pin::new(b), cx, buf),
-        }
-    }
-}
-
-// TODO: remove?
-impl<A, B> Read for EitherOutput<A, B>
-where
-    A: Read,
-    B: Read,
-{
-    fn read(&mut self, buf: &mut [u8]) -> Result<usize, IoError> {
-        match self {
-            EitherOutput::First(a) => a.read(buf),
-            EitherOutput::Second(b) => b.read(buf),
+    #[project]
+    fn poll_read(self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, IoError>> {
+        #[project]
+        match self.project() {
+            EitherOutput::First(a) => AsyncRead::poll_read(a, cx, buf),
```

Comment:
Here and elsewhere you switched to UFCS style. What is the motivation for this change and what is the advantage here over writing `a.poll_read(cx, buf)`?

Reply:
It's just a habit from early days of `futures-preview` where it wasn't possible to write `a.poll_read`.